dspy/signatures/signature.py

```diff
@@ -34,22 +34,75 @@ def __new__(cls, name, bases, class_dict):
         return new_class
 
     def __call__(cls, *args, **kwargs):
+        if len(args) == 1 and isinstance(args[0], str):
+            instance = super(SignatureMeta, cls).__call__(*args, **kwargs)
+            return instance
+        #old 
         return cls._template(*args, **kwargs)
 
     def __getattr__(cls, attr):
         # Redirect attribute access to the template object when accessed on the class directly
         return getattr(cls._template, attr)
+
+
+class Signature(metaclass=SignatureMeta):
+    def __init__(self, signature: str = "", instructions: str = ""):
+        self.signature = signature
+        self.instructions = instructions
+        self.fields = {}
+        self.parse_structure()
     
-    def __repr__(cls):
-        s = []
+    @property
+    def kwargs(self):
+        return {k: v for k, v in self.fields.items()}
+
+    def parse_structure(self):
+        inputs_str, outputs_str = self.signature.split("->")
+        for name in inputs_str.split(","):
+            self.add_field(name.strip(), InputField())
+        for name in outputs_str.split(","):
+            self.add_field(name.strip(), OutputField())
+
+    def attach(self, **kwargs):
+        for key, (prefix, desc) in kwargs.items():
+            field_type = self.fields.get(key)
+            if not field_type:
+                raise ValueError(f"{key} does not exist in this signature")
+            field_map = {
+                InputField: InputField(prefix=prefix, desc=desc),
+                OutputField: OutputField(prefix=prefix, desc=desc)
+            }
+            self.fields[key] = field_map.get(type(field_type))
+        return self
+
+    def add_field(self, field_name: str, field_type, position="append"):
+        if field_name in self.fields:
+            raise ValueError(f"{field_name} already exists in fields.")
+        if isinstance(field_type, (InputField, OutputField)):
+            field_instance = field_type
+        else:
+            raise ValueError(f"non-existent {field_type}.")
+        if position == "prepend":
```

Comment:
I don't think this will handle append/prepend correctly considering that inputs must all go before outputs

Reply:
Probably need to find the endpoint of the input fields (for append to inputs or prepend to outputs)